File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>

using namespace std;

int k;
string s;

int main() {
	
    ios::sync_with_stdio(false);
    cin.tie(NULL);
	
	cin >> k;
	
	for(int i=0;i<k;i++){
		cin >> s;
		stack<char> stk;

		for(int j=0;j<s.length();j++){
			if(!stk.empty() && stk.top()=='(' && s[j]== ')') stk.pop();
			else stk.push(s[j]);
		}
		if(stk.size()==0) cout << "YES" << "\n";
		else cout << "NO" << "\n";
	}

	return 0;	
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys

a = int(sys.stdin.readline())


for i in range(0,a):
    k = sys.stdin.readline().strip()
    stack = []
    for c in k:
        if len(stack) == 0:
            stack.append(c)
        elif stack and stack[-1] == '(' and c == ')':
            stack.pop()
        else :
            stack.append(c)

    if len(stack)==0 :
        print("YES")
    else :
        print("NO")
